File: code/141.linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None
import ll

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        slow, fast = head;
        while slow != None:
            slow = slow.next
            fast = fast.next.next

            if slow == fast:
                return True

        return False
        
        
def main():
    

main()

File: code/rust/141.linked-list-cycle.rs
// mod ll;
// use ll::*;
impl Solution {
    pub fn has_circle(mut head: Option<Box<ListNode>>) -> Option<bool> {
        let mut slow = &mut head;
        let mut fast = slow.clone();

        while slow.is_some() && slow.as_ref()?.next.as_ref()?.next.is_some() {
            slow = &mut slow.as_mut()?.next;
            fast = fast.unwrap().next.unwrap().next;

            if slow.as_ref()? == fast.as_ref()? {
                return Some(true);
            }
        }

        return None;
    }
}

// fn main() {
//     let ln = ListNode::from(vec![1, 2, 3, 5, 3, 5]);
//     let r = Solution::has_circle(ln);
//     println!("{:?}", r);
// }
